### DISCUSSION TITLE
Quick particle question - what is the correct way to add an emitter to an actor?

### DISCUSSION THREAD

**Discussion by chrisk-7777** [excaliburjs/Excalibur#2433] [Q&A] [answered]
Probably a simple question, but my smooth brain isn't working atm:

## Works

```
const game = new ex.Engine({
  width: 400,
  height: 400
});

const emitter = new ex.ParticleEmitter({
  width: 10,
  height: 10,
  x: 10,
  y: 10,
  emitterType: ex.EmitterType.Rectangle,
  radius: 5,
  minVel: 100,
  maxVel: 200,
  minAngle: 0,
  maxAngle: 6.2,
  emitRate: 300,
  opacity: 0.5,
  fadeFlag: true,
  particleLife: 1000,
  maxSize: 10,
  minSize: 1,
  startSize: 5,
  endSize: 100,
  acceleration: ex.vec(10, 80),
  beginColor: ex.Color.Chartreuse,
  endColor: ex.Color.Magenta
});

game.start();
game.add(emitter)
```

## Does not work

```
const game = new ex.Engine({
  width: 400,
  height: 400
});

const actor = new ex.Actor({
  anchor: ex.vec(0.5, 0.5),
  pos: game.screen.center,
  color: ex.Color.Red,
  radius: 5,
});

const emitter = new ex.ParticleEmitter({
  width: 10,
  height: 10,
  x: 10,
  y: 10,
  emitterType: ex.EmitterType.Rectangle,
  radius: 5,
  minVel: 100,
  maxVel: 200,
  minAngle: 0,
  maxAngle: 6.2,
  emitRate: 300,
  opacity: 0.5,
  fadeFlag: true,
  particleLife: 1000,
  maxSize: 10,
  minSize: 1,
  startSize: 5,
  endSize: 100,
  acceleration: ex.vec(10, 80),
  beginColor: ex.Color.Chartreuse,
  endColor: ex.Color.Magenta
});

game.start();
game.add(actor);
actor.addChild(emitter);

```

Link: https://codepen.io/chrisk7777/pen/poLroev?editors=0010

I note in the docs that it refers to `actor.add()` , but I think that might be out of date? So I'm taking a stab at `addChild`
https://excaliburjs.com/docs/particles

I'm happy to open a doc PR once its all sussed out. 

## Side question

I noticed in one of the issues that using the `particleSprite` may not be a good idea. Is that still the case?

**Comment by eonarheim** [comment, marked as answer]
@chrisk-7777 In theory adding an emitter as a child to an entity should work, very possibly a bug with particle emitters... I'm looking into it.

I think the docs are out of date for sure 👍 

I think it's fine to use `particleSprite` now, the reason that message was there was in the past sprite modifications were expensive to animate. I think that's some stale information/documentation. Apologies, the particle emitter api is definitely one of the oldest in Excalibur and needs some work, it's definitely not the most friendly.

**Reply by eonarheim** [reply]
Can confirm, the `ParticleEmitter` has at least 2 bugs:
1. #2430 fixes an issue in the latest alpha, which prevents the `ParticleEmitter` from adding particles to the current scene in v0.27.0
    <img width="377" alt="image" src="https://user-images.githubusercontent.com/612071/180667959-206f5975-ca91-4070-92d7-d00ef832768f.png">
   
    a. To work around this in the non-alpha you could manually add the scene to the emitter `emitter.scene = game.currentScene`

2. There is another bug where it doesn't respect the parented transform when generating particles, so it will only generate particles in world space as if there was no parent entity. There is a possible workaround by setting the emitter position to be based on the parent.
    ```
    emitter.onPreUpdate = () => {
      emitter.pos = actor.pos;
    }
    ```

**Reply by chrisk-7777** [reply]
That's fantastic, works perfectly. I'll hold off until this alpha makes it to a release and I'll put together some docs (this also gives me time to get my head around it)

Updated codepen: https://codepen.io/chrisk7777/pen/poLroev?editors=0010

Thanks as always!

![CleanShot 2022-07-26 at 00 16 41](https://user-images.githubusercontent.com/55621012/180798913-aa16fcb8-646f-4a46-9d3a-278811b5433d.gif)

**Reply by eonarheim** [reply]
Awesome this looks super cool!

Almost forgot, but here is the bug https://github.com/excaliburjs/Excalibur/issues/2437

**Reply by chrisk-7777** [reply]
Its amazing what you can with a few lines in excaliburjs. 
I'm still getting my head around the sprite emitter, but I'll loop back to that in a bit 

Thanks for the link to the bug
